src/workerd/api/queue.c++

```diff
@@ -24,18 +24,34 @@ kj::StringPtr validateContentType(kj::StringPtr contentType) {
   }
 }
 
-kj::Array<kj::byte> serializeV8(jsg::Lock& js, v8::Local<v8::Value> body) {
+struct Serialized {
+  kj::Maybe<kj::OneOf<kj::String, kj::Array<kj::byte>, jsg::BufferSource>> own;
+  // Holds onto the owner of a given array of serialized data.
+  kj::ArrayPtr<kj::byte> data;
+  // A pointer into that data that can be directly written into an outgoing queue send, regardless
+  // of its holder.
+};
+
+Serialized serializeV8(jsg::Lock& js, v8::Local<v8::Value> body) {
   // Use a specific serialization version to avoid sending messages using a new version before all
   // runtimes at the edge know how to read it.
   jsg::Serializer serializer(js.v8Isolate, jsg::Serializer::Options {
     .version = 15,
     .omitHeader = false,
   });
   serializer.write(body);
-  return serializer.release().data;
+  kj::Array<kj::byte> bytes = serializer.release().data;
+  Serialized result;
+  result.data = bytes;
+  result.own = kj::mv(bytes);
+  return kj::mv(result);
 }
 
-kj::Array<kj::byte> serialize(jsg::Lock& js, v8::Local<v8::Value> body, kj::StringPtr contentType) {
+Serialized serialize(
+    jsg::Lock& js,
+    v8::Local<v8::Value> body,
+    kj::StringPtr contentType,
+    bool forSynchronousUse = false) {
```

Comment:
nit: generally not a fan of bool args like this as it makes the call site a bit less obvious. Would prefer an enum here instead but not blocking.

Reply:
Yeah, that's pretty much always clearer, and it's fair to call me out on it. Updated.